### DISCUSSION TITLE
AST Transformations

### DISCUSSION THREAD

**Discussion by will62794** [tlaplus-community/tree-sitter-tlaplus#64] [Q&A]
Do you know if there is there any mechanism for describing transformations on the parsed tree-sitter AST in some standardized way? I've been sort of doing an ad hoc version of this in the JS TLA+ interpreter by replacing AST nodes in their corresponding locations in the source text and then re-parsing. Just curious if there were any existing facilities for this built into the tree-sitter API.

**Comment by ahelwer** [comment]
Not that I know of. Outputting the AST as an S-expression then using a scheme-like language such as Racket to transform it could be an idea, but might not work well given the JavaScript constraint. Are you talking about using the AST to rewrite the actual TLA+ code, or just rewriting the AST itself?

**Reply by will62794** [reply]
Basically just re-writing the AST itself e.g. for implementing various syntactic pre-processing steps (e.g. de-sugaring, expanding constants, definitions, etc.) to simplify certain aspects of expression evaluation. I'm doing some [hacky stuff](https://github.com/will62794/tla-web/blob/224cd80cfa2bfb7d447f7b929fd4d1fc260c7df5/js/eval.js#L567-L582) to emulate this right now :)

**Reply by ahelwer** [reply]
Ah, I see. There might be some tree-sitter based refactoring projects kicking around you could look at, but you're ahead of me in the interpreter game so I don't know too much here. You could also ask this question in the discussions tab of tree-sitter itself.

**Reply by will62794** [reply]
Ok, thanks!
